### PR TITLE
raftstore: support to output old value to CmdObserver

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

### What problem does this PR solve?

Problem Summary:
Old Value is the previously written value before a delete/update operation.
To support the new collation format and compatible with TiDB Binlog, TiCDC requires TiKV to output the old value for it.

### What is changed and how it works?
Separate from #8071

What's Changed:

* Add a new method `RaftstoreRouter::send_command_txn_extra`
* Add a cache for `TxnExtra` in `ApplyContext`
* Send `TxnExtra` through `CmdObserver`

### Related changes

- Should be cherry-picked to release 4.0 after all related PRs got merged.

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 
* N/A

### CODE REVIEW

**Comment by 5kbpers** [comment]
PTAL @BusyJay @Little-Wallace @gengliqi @overvenus

File: components/raftstore/src/store/peer.rs [hunk 1]
```diff
@@ -277,6 +280,8 @@ pub struct Peer {
     /// Send to these peers to check whether itself is stale.
     pub check_stale_conf_ver: u64,
     pub check_stale_peers: Vec<metapb::Peer>,
+
```
Comment (gengliqi) [inline components/raftstore/src/store/peer.rs:283]: How about adding an annotation?

**Pull request merged**